File: packages/python-sdk/src/relay_sdk/ws.py
"""WebSocket client for real-time Relay events."""

from __future__ import annotations

import asyncio
import json
import logging
from typing import Any, Callable
from urllib.parse import quote

import websockets
from websockets.asyncio.client import ClientConnection

logger = logging.getLogger(__name__)

EventHandler = Callable[[dict[str, Any]], None]

_MAX_RECONNECT_ATTEMPTS = 10
_PING_INTERVAL_SECONDS = 30


class WsClient:
    """Async WebSocket client with auto-reconnection and event handling.

    Usage::

        ws = WsClient(token="at_xxx")
        ws.on("message.created", lambda e: print(e))
        ws.subscribe(["general", "code-review"])
        await ws.connect()  # blocks until disconnect() is called
    """

    def __init__(self, token: str, *, base_url: str | None = None) -> None:
        self._token = token
        base = (base_url or "https://api.agentrelay.dev").rstrip("/")
        self._base_url = base.replace("https://", "wss://").replace("http://", "ws://")
        self._ws: ClientConnection | None = None
        self._handlers: dict[str, set[EventHandler]] = {}
        self._reconnect_attempt = 0
        self._closed = False
        self._ping_task: asyncio.Task[None] | None = None
        self._pending_subscriptions: list[str] = []

    async def connect(self) -> None:
        """Connect to the WebSocket and process events until disconnected."""
        self._closed = False
        while not self._closed:
            try:
                await self._connect_once()
            except Exception:
                if self._closed:
                    break
                if self._reconnect_attempt >= _MAX_RECONNECT_ATTEMPTS:
                    logger.error("Max reconnect attempts reached")
                    break
                delay = min(2**self._reconnect_attempt, 30)
                self._reconnect_attempt += 1
                logger.info("Reconnecting in %ds (attempt %d)", delay, self._reconnect_attempt)
                await asyncio.sleep(delay)

    async def _connect_once(self) -> None:
        url = f"{self._base_url}/v1/stream?token={quote(self._token, safe='')}"
        async with websockets.connect(url) as ws:
            self._ws = ws
            self._reconnect_attempt = 0
            self._start_ping()
            self._emit("open", {"type": "open"})

            # Re-subscribe to any pending channels
            if self._pending_subscriptions:
                await self._send_json({"type": "subscribe", "channels": self._pending_subscriptions})

            try:
                async for raw in ws:
                    try:
                        data = json.loads(raw)
                        self._emit(data.get("type", ""), data)
                    except (json.JSONDecodeError, TypeError):
                        pass
            finally:
                self._stop_ping()
                self._ws = None
                self._emit("close", {"type": "close"})

    def disconnect(self) -> None:
        """Signal disconnect. The connect() loop will exit."""
        self._closed = True
        self._stop_ping()
        if self._ws:
            asyncio.ensure_future(self._ws.close())

    def subscribe(self, channels: list[str]) -> None:
        """Subscribe to channels. If not connected yet, subscribes on connect."""
        self._pending_subscriptions = list(set(self._pending_subscriptions + channels))
        if self._ws:
            asyncio.ensure_future(self._send_json({"type": "subscribe", "channels": channels}))

    def unsubscribe(self, channels: list[str]) -> None:
        """Unsubscribe from channels."""
        self._pending_subscriptions = [c for c in self._pending_subscriptions if c not in channels]
        if self._ws:
            asyncio.ensure_future(self._send_json({"type": "unsubscribe", "channels": channels}))

    def on(self, event: str, handler: EventHandler) -> Callable[[], None]:
        """Register an event handler. Returns an unsubscribe callable."""
        if event not in self._handlers:
            self._handlers[event] = set()
        self._handlers[event].add(handler)

        def remove() -> None:
            self._handlers.get(event, set()).discard(handler)

        return remove

    def off(self, event: str, handler: EventHandler) -> None:
        """Remove an event handler."""
        self._handlers.get(event, set()).discard(handler)

    def _emit(self, event: str, data: dict[str, Any]) -> None:
        for h in self._handlers.get(event, set()).copy():
            try:
                h(data)
            except Exception:
                logger.exception("Event handler error for %s", event)
        for h in self._handlers.get("*", set()).copy():
            try:
                h(data)
            except Exception:
                logger.exception("Wildcard handler error for %s", event)

    async def _send_json(self, data: Any) -> None:
        if self._ws:
            try:
                await self._ws.send(json.dumps(data))
            except Exception:
                pass

    def _start_ping(self) -> None:
        self._stop_ping()
        self._ping_task = asyncio.ensure_future(self._ping_loop())

    def _stop_ping(self) -> None:
        if self._ping_task:
            self._ping_task.cancel()
            self._ping_task = None

    async def _ping_loop(self) -> None:
        try:
            while True:
                await asyncio.sleep(_PING_INTERVAL_SECONDS)
                await self._send_json({"type": "ping"})
        except asyncio.CancelledError:
            pass

File: packages/sdk/src/ws.ts
import type { WsClientEvent, WsOpenEvent, WsErrorEvent, WsReconnectingEvent, WsCloseEvent } from '@relaycast/types';
import { ServerEventSchema } from '@relaycast/types';

export type EventHandler<T = WsClientEvent> = (event: T) => void;

export interface WsClientOptions {
  token: string;
  baseUrl?: string;
  /** Log warnings for dropped/malformed WebSocket messages via console.warn. */
  debug?: boolean;
}

export class WsClient {
  private token: string;
  private baseUrl: string;
  private ws: WebSocket | null = null;
  private handlers: Map<string, Set<EventHandler>> = new Map();
  private reconnectAttempt = 0;
  private maxReconnectAttempts = 10;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private pingTimer: ReturnType<typeof setInterval> | null = null;
  private closed = false;
  private debug: boolean;

  constructor(options: WsClientOptions) {
    this.token = options.token;
    this.debug = options.debug ?? false;
    const base = options.baseUrl ?? 'https://api.relaycast.dev';
    this.baseUrl = base.replace(/^http/, 'ws');
  }

  connect(): void {
    if (this.ws) return;
    this.closed = false;

    const url = `${this.baseUrl}/v1/ws?token=${encodeURIComponent(this.token)}`;
    this.ws = new WebSocket(url);

    this.ws.onopen = () => {
      this.reconnectAttempt = 0;
      this.startPing();
      const openEvent: WsOpenEvent = { type: 'open' };
      this.emit('open', openEvent);
    };

    this.ws.onmessage = (event: MessageEvent) => {
      try {
        const parsed = JSON.parse(String(event.data));
        const result = ServerEventSchema.safeParse(parsed);
        if (result.success) {
          this.emit(result.data.type, result.data);
        } else if (
          parsed !== null &&
          typeof parsed === 'object' &&
          typeof parsed.type === 'string'
        ) {
          // Forward unrecognized event types so wildcard listeners still fire
          this.emit(parsed.type, parsed as WsClientEvent);
        } else if (this.debug) {
          console.warn('[relaycast] Dropped WebSocket message: missing or invalid "type" field', parsed);
        }
      } catch (err) {
        if (this.debug) {
          console.warn('[relaycast] Dropped malformed WebSocket message:', String(event.data).slice(0, 200), err);
        }
      }
    };

    this.ws.onclose = () => {
      this.stopPing();
      this.ws = null;
      if (!this.closed) {
        this.scheduleReconnect();
      }
      const closeEvent: WsCloseEvent = { type: 'close' };
      this.emit('close', closeEvent);
    };

    this.ws.onerror = () => {
      const errorEvent: WsErrorEvent = { type: 'error' };
      this.emit('error', errorEvent);
    };
  }

  disconnect(): void {
    this.closed = true;
    this.stopPing();
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  subscribe(channels: string[]): void {
    this.sendJson({ type: 'subscribe', channels });
  }

  unsubscribe(channels: string[]): void {
    this.sendJson({ type: 'unsubscribe', channels });
  }

  on(event: string, handler: EventHandler): () => void {
    if (!this.handlers.has(event)) {
      this.handlers.set(event, new Set());
    }
    this.handlers.get(event)!.add(handler);
    return () => {
      this.handlers.get(event)?.delete(handler);
    };
  }

  off(event: string, handler: EventHandler): void {
    this.handlers.get(event)?.delete(handler);
  }

  private emit(event: string, data: WsClientEvent): void {
    this.handlers.get(event)?.forEach((h) => h(data));
    this.handlers.get('*')?.forEach((h) => h(data));
  }

  private sendJson(data: unknown): void {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(data));
    }
  }

  private startPing(): void {
    this.stopPing();
    this.pingTimer = setInterval(() => {
      this.sendJson({ type: 'ping' });
    }, 30_000);
  }

  private stopPing(): void {
    if (this.pingTimer) {
      clearInterval(this.pingTimer);
      this.pingTimer = null;
    }
  }

  private scheduleReconnect(): void {
    if (this.reconnectAttempt >= this.maxReconnectAttempts) return;
    const delay = Math.min(1000 * Math.pow(2, this.reconnectAttempt), 30_000);
    this.reconnectAttempt++;
    const reconnectingEvent: WsReconnectingEvent = { type: 'reconnecting', attempt: this.reconnectAttempt };
    this.emit('reconnecting', reconnectingEvent);
    this.reconnectTimer = setTimeout(() => {
      this.connect();
    }, delay);
  }
}
